### PR TITLE
add is debugger paused to task api

### PR DESCRIPTION
**Pull request by sofianhnaide**
## Why are these changes needed?
This change adds a flag `is_debugger_paused` to the task API. It allows querying by that flag to find tasks that has it set to true. (unit test are coming shortly)

### Manual testing

I ran the following script.

```
import ray

@ray.remote
def test2():
    with ray._private.worker.global_worker.task_paused_by_debugger():
        print("waiting")
        import time
        time.sleep(60)

    print("changing to false")
    time.sleep(60)

result = test2.remote()
print(ray.get(result))
```

and then queried ray by running `ray list tasks -f "is_debugger_paused=True"` and `ray list tasks -f "is_debugger_paused=False"`

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/_private/worker.py
Comment (rickyyx) [inline python/ray/_private/worker.py:564]: is this just for testing?

File: python/ray/util/state/common.py
Comment (rickyyx) [inline python/ray/util/state/common.py:747]: nit: i wonder if we can add more metadata about the debugging? e.g. do we know where this task is stopped, at which port or etc, other than just a boolean? Would that be possible?

File: python/ray/util/state/state_manager.py
Comment (rickyyx) [inline python/ray/util/state/state_manager.py:311]: oh so we didn't have a `bool` SupportedFilterType here? I actually wonder if this filter needs to be pushed down to GCS if we can just do the filtering in `_filter` in `dashboard/state_aggregator.py` after listing everything from GCS, so we can rely on the boolean filter here: 

https://github.com/ray-project/ray/blob/20f319c2b7642981363093beef3b478c0eb5d052/dashboard/state_aggregator.py#L124-L133

**Review by rickyyx — COMMENTED** [review]
Looks pretty good to me.

File: python/ray/_private/worker.py
Comment (sofianhnaide) [reply python/ray/_private/worker.py:564]: this is the api that can be used from other parts of ray by doing:
```
 with ray._private.worker.global_worker.task_paused_by_debugger():
     # setup the debugger and wait for the client
```

File: python/ray/util/state/common.py
Comment (sofianhnaide) [reply python/ray/util/state/common.py:747]: The only other part that we need is the port and that is already set on the worker, for now the behavior we need is to list tasks filtered by this flag, and only after the user select the task we can then get the worker to get the ip, and the debugger port. So we really do have everything we need here.

File: python/ray/util/state/state_manager.py
Comment (sofianhnaide) [reply python/ray/util/state/state_manager.py]: Actually, looking through it, this is not needed. I added it when I was debugging why filtering did not work but I removed it now and it still works.

This is now removed.

File: src/ray/core_worker/core_worker.cc
Comment (rkooo567) [inline src/ray/core_worker/core_worker.cc:4338]: I think this may not work with async actor (it is due to nasty tech debt in ray core). There's a hacky workaround we can use, and we may need to use the same workaround for this.

Can you try adding unit tests and see if it works with async actor / threaded actor / regular actor? 

- async actor: actor with async method
- threaded actor: actor with num_concurrency > 1 inside ray.remote
- regular actor: the actor you know

If it turns out async actor doesn't work (which I suspect), let me give you a code pointer here.

**Review by rkooo567 — COMMENTED** [review]
Code looks good! I think it may not work with async actor (due to ray core's tech debt). We can discuss how to enable it in that case after adding unit tests

**Comment by sofianhnaide** [comment]
> Code looks good! I think it may not work with async actor (due to ray core's tech debt). We can discuss how to enable it in that case after adding unit tests

@rkooo567 i just added unit tests, what's the case that wouldn't work for Async actors?

File: src/ray/core_worker/core_worker.cc
Comment (sofianhnaide) [reply src/ray/core_worker/core_worker.cc:4338]: Can you point me to unit tests that test similar logic. I looked at core_worker_test but didn't find anything relevant.

**Comment by rkooo567** [comment]
i think you can see

test_task_event.py

```
test_parent_task_id_non_concurrent_actor
test_parent_task_id_concurrent_actor
```

Basically what you need to check is

1. For a normal actor, we probably don't need e2e tests, but nice to add it here.
2. For a threaded actor, execute 2 tasks concurrently (just call actor.task.remote() twice with num_concurrency=2, and each task just sleeps long time). Add breakpoint to one of tasks and make sure only one task has paused = true
3. For async actor, I think you can trigger the edge case this way;

```python3

class AsyncActor:
    async def f(self, pause):
        await asyncio.sleep(2)
        if pause:
            breakpoint()
        else:
            await asyncio.sleep(10)

a = AsyncActor.remote()
ref1 = a.f.remote(True)
time.sleep(0.5)
ref2 = a.f.remote(False)
# Check if the first task pause = True
```

It is because whenever new task is submitted, it overwrites the "current_task_id" from the worker context to the newly submitted task (it is a bug). And your task event always finds the task id from the current task from worker context

File: python/ray/_private/worker.py
Comment (rickyyx) [reply python/ray/_private/worker.py:564]: i see - so this is aimed as a developer API? I guess I am not too sure when "other parts of ray" would be doing this other than during testing. Anything you have in mind?

**Comment by rickyyx** [comment]
> It is because whenever new task is submitted, it overwrites the "current_task_id" from the worker context to the newly submitted task (it is a bug)

I thought it's the task id for "executing task" that matters here tho? Why would task submission affects current task id. @rkooo567 wanna sync offline for this? I feel this is ultimately the same issue we ran into with task logs for async actors. 

> Can you point me to the work around? I will also add the E2E tests.

My current intuition is that we will need to pass the task id in (rather than getting it from the worker context which is not synchronized with async actor's task switching since async tasks run in the same thread, thus the same worker context since it's now thread local).

**Comment by sofianhnaide** [comment]
> My current intuition is that we will need to pass the task id in (rather than getting it from the worker context which is not synchronized with async actor's task switching since async tasks run in the same thread, thus the same worker context since it's now thread local).  
> 
> 

That works, I actually thought of that and tried looking for any example of getting the task Id from the python land to pass it but I couldn't find it.

**Comment by rkooo567** [comment]
Currently, you can access the task id from 

```
# This argument is used to obtain the correct task id inside
# an asyncio task. It is because task_id can be obtained
# by the worker_context_ API, which is per thread, not per
# asyncio task. TODO(sang): We should properly fix it.
# Note that the context var is recommended to be defined
# in the top module.
# https://docs.python.org/3/library/contextvars.html#contextvars.ContextVar
# It is thread-safe.
async_task_id = contextvars.ContextVar('async_task_id', default=None)
```

this inside raylet.pyx

**Comment by rkooo567** [comment]
I think we can make the API go through raylet.pyx and always manually pass task_id

**Comment by rkooo567** [comment]
So if async -> use async context var

if not -> use worker context (it should work with both threaded/normal actor)

File: dashboard/state_aggregator.py
Comment (rickyyx) [inline dashboard/state_aggregator.py:193]: I thought this should already be covered by the `_convert_filters_type` above? Or we need the manual matching here?

File: python/ray/_private/worker.py
Comment (sofianhnaide) [reply python/ray/_private/worker.py]: what we have in mind is something similar to `rpdb.py`, you can check the full prototype PR for more info

File: python/ray/tests/test_task_events.py
Comment (rickyyx) [inline python/ray/tests/test_task_events.py:534]: nit: should we do 1 and 3?

File: src/ray/core_worker/core_worker.cc
Comment (rickyyx) [inline src/ray/core_worker/core_worker.cc:4341]: let's do `rpc::TaskStatus::NIL` since we are not updating task status for debugger?

File: src/ray/protobuf/gcs.proto
Comment (rickyyx) [inline src/ray/protobuf/gcs.proto:248]: we don't need to pass this as GCS's filter since we are no longer pushing down the filter for debugger to GCS?

**Review by rickyyx — COMMENTED** [review]
Great testing! 

Some minor comments from I believe previous clean-ups.

File: src/ray/core_worker/core_worker.cc
Comment (sofianhnaide) [reply src/ray/core_worker/core_worker.cc:4341]: I don't think I fully understand how `RecordTaskStatusEvent` works so I will leave that up to you. Is replacing it with nil will ignore updating the state or reset it?

File: src/ray/protobuf/gcs.proto
Comment (sofianhnaide) [reply src/ray/protobuf/gcs.proto:248]: I think we are here right? https://github.com/ray-project/ray/pull/41067/files#diff-619b019a6fa687fece5fdb2058396e7e959c39010af5c3b027208fa4b5cc7eaeR106

File: dashboard/state_aggregator.py
Comment (sofianhnaide) [reply dashboard/state_aggregator.py:193]: from my testing, we do need it here.

File: python/ray/tests/test_task_events.py
Comment (sofianhnaide) [reply python/ray/tests/test_task_events.py]: done

File: src/ray/core_worker/core_worker.cc
Comment (sofianhnaide) [reply src/ray/core_worker/core_worker.cc]: Alright did some testing, replacing with nil seems harmless. I updated it.

File: dashboard/state_aggregator.py
Comment (rickyyx) [reply dashboard/state_aggregator.py:193]: ah i think I know why, we have a bug in matching Optional column. Added an issue to track: https://github.com/ray-project/ray/issues/41132

File: src/ray/protobuf/gcs.proto
Comment (rickyyx) [reply src/ray/protobuf/gcs.proto:248]: sorry - i meant another place.

File: src/ray/protobuf/gcs_service.proto
Comment (rickyyx) [inline src/ray/protobuf/gcs_service.proto:671]: is this needed since we are not pushing down filters to GCS?

Comment (sofianhnaide) [reply src/ray/protobuf/gcs_service.proto:671]: yes, I use it here: https://github.com/ray-project/ray/pull/41067/files#diff-bb7747741306d767d5a35c0e55cde709d54cc407b24c806a61d63392795e1303R350

Comment (rickyyx) [reply src/ray/protobuf/gcs_service.proto:671]: like on the caller side, we don't have code that sets it anymore? I beleive it was needed in earlier iterations here: https://github.com/ray-project/ray/pull/41067#discussion_r1388743316

Comment (sofianhnaide) [reply src/ray/protobuf/gcs_service.proto]: got it, removed.

**Review by rkooo567 — APPROVED** [review]
I am okay to merge this once @rickyyx 's reviews are addressed!

File: src/ray/gcs/gcs_server/test/gcs_task_manager_test.cc
Comment (rickyyx) [inline src/ray/gcs/gcs_server/test/gcs_task_manager_test.cc:149]: nit: this test shouldn't be passing now?

**Review by rickyyx — APPROVED** [review]
Thanks for the great PR!

**Pull request merged**
